// react-table/lib/methods.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

exports.default = {
  getDataModel: function getDataModel(newState) {
    var _this = this;

    var columns = newState.columns,
        _newState$pivotBy = newState.pivotBy,
        pivotBy = _newState$pivotBy === undefined ? [] : _newState$pivotBy,
        data = newState.data,
        pivotIDKey = newState.pivotIDKey,
        pivotValKey = newState.pivotValKey,
        subRowsKey = newState.subRowsKey,
        expanderColumnWidth = newState.expanderColumnWidth,
        SubComponent = newState.SubComponent;

    // Determine Header Groups

    var hasHeaderGroups = false;
    columns.forEach(function (column) {
      if (column.columns) {
        hasHeaderGroups = true;
      }
    });

    // Build Header Groups
    var headerGroups = [];
    var currentSpan = [];

    // A convenience function to add a header and reset the currentSpan
    var addHeader = function addHeader(columns) {
      var column = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : columns[0];

      headerGroups.push(_extends({}, _this.props.column, column, {
        columns: columns
      }));
      currentSpan = [];
    };

    var noSubExpanderColumns = columns.map(function (col) {
      return _extends({}, col, {
        columns: col.columns ? col.columns.filter(function (d) {
          return !d.expander;
        }) : undefined
      });
    });

    var expanderColumnIndex = columns.findIndex(function (col) {
      return col.expander;
    });
    var needsExpander = (SubComponent || pivotBy.length) && expanderColumnIndex === -1;
    var columnsWithExpander = needsExpander ? [{ expander: true }].concat(_toConsumableArray(noSubExpanderColumns)) : noSubExpanderColumns;
    if (needsExpander) {
      expanderColumnIndex = 0;
    }

    var makeDecoratedColumn = function makeDecoratedColumn(column) {
      var dcol = _extends({}, _this.props.column, column);

      if (dcol.expander) {
        dcol.width = expanderColumnWidth;
        return dcol;
      }

      if (typeof dcol.accessor === 'string') {
        var _ret = function () {
          dcol.id = dcol.id || dcol.accessor;
          var accessorString = dcol.accessor;
          dcol.accessor = function (row) {
            return _utils2.default.get(row, accessorString);
          };
          return {
            v: dcol
          };
        }();

        if ((typeof _ret === 'undefined' ? 'undefined' : _typeof(_ret)) === "object") return _ret.v;
      }

      if (dcol.accessor && !dcol.id) {
        console.warn(dcol);
        throw new Error('A column id is required if using a non-string accessor for column above.');
      }

      if (!dcol.accessor) {
        dcol.accessor = function (d) {
          return undefined;
        };
      }

      // Ensure minWidth is not greater than maxWidth if set
      if (dcol.maxWidth < dcol.minWidth) {
        dcol.minWidth = dcol.maxWidth;
      }

      return dcol;
    };

    // Decorate the columns
    var decorateAndAddToAll = function decorateAndAddToAll(col) {
      var decoratedColumn = makeDecoratedColumn(col);
      allDecoratedColumns.push(decoratedColumn);
      return decoratedColumn;
    };
    var allDecoratedColumns = [];
    var decoratedColumns = columnsWithExpander.map(function (column, i) {
      if (column.columns) {
        return _extends({}, column, {
          columns: column.columns.map(decorateAndAddToAll)
        });
      } else {
        return decorateAndAddToAll(column);
      }
    });

    // Build the visible columns, headers and flat column list
    var visibleColumns = decoratedColumns.slice();
    var allVisibleColumns = [];

    visibleColumns = visibleColumns.map(function (column, i) {
      if (column.columns) {
        var visibleSubColumns = column.columns.filter(function (d) {
          return pivotBy.indexOf(d.id) > -1 ? false : _utils2.default.getFirstDefined(d.show, true);
        });
        return _extends({}, column, {
          columns: visibleSubColumns
        });
      }
      return column;
    });

    visibleColumns = visibleColumns.filter(function (column) {
      return column.columns ? column.columns.length : pivotBy.indexOf(column.id) > -1 ? false : _utils2.default.getFirstDefined(column.show, true);
    });

    // Move the pivot columns into a single column if needed
    if (pivotBy.length) {
      var pivotColumns = [];
      for (var i = 0; i < allDecoratedColumns.length; i++) {
        if (pivotBy.indexOf(allDecoratedColumns[i].id) > -1) {
          pivotColumns.push(allDecoratedColumns[i]);
        }
      }
      var _pivotColumn = _extends({}, pivotColumns[0], {
        pivotColumns: pivotColumns,
        expander: true
      });
      visibleColumns[expanderColumnIndex] = _pivotColumn;
    }

    // Build flast list of allVisibleColumns and HeaderGroups
    visibleColumns.forEach(function (column, i) {
      if (column.columns) {
        allVisibleColumns = allVisibleColumns.concat(column.columns);
        if (currentSpan.length > 0) {
          addHeader(currentSpan);
        }
        addHeader(column.columns, column);
        return;
      }
      allVisibleColumns.push(column);
      currentSpan.push(column);
    });
    if (hasHeaderGroups && currentSpan.length > 0) {
      addHeader(currentSpan);
    }

    // Access the data
    var resolvedData = data.map(function (d, i) {
      var row = {
        __original: d,
        __index: i
      };
      allDecoratedColumns.forEach(function (column) {
        if (column.expander) return;
        row[column.id] = column.accessor(d);
      });
      return row;
    });

    // If pivoting, recursively group the data
    var aggregate = function aggregate(rows) {
      var aggregationValues = {};
      aggregatingColumns.forEach(function (column) {
        var values = rows.map(function (d) {
          return d[column.id];
        });
        aggregationValues[column.id] = column.aggregate(values, rows);
      });
      return aggregationValues;
    };
    var standardColumns = pivotBy.length ? allVisibleColumns.slice(1) : allVisibleColumns;
    var aggregatingColumns = standardColumns.filter(function (d) {
      return d.aggregate;
    });
    var pivotColumn = void 0;
    if (pivotBy.length) {
      (function () {
        pivotColumn = allVisibleColumns[0];
        var groupRecursively = function groupRecursively(rows, keys) {
          var i = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

          // This is the last level, just return the rows
          if (i === keys.length) {
            return rows;
          }
          // Group the rows together for this level
          var groupedRows = Object.entries(_utils2.default.groupBy(rows, keys[i])).map(function (_ref) {
            var _ref3;

            var _ref2 = _slicedToArray(_ref, 2),
                key = _ref2[0],
                value = _ref2[1];

            return _ref3 = {}, _defineProperty(_ref3, pivotIDKey, keys[i]), _defineProperty(_ref3, pivotValKey, key), _defineProperty(_ref3, keys[i], key), _defineProperty(_ref3, subRowsKey, value), _ref3;
          });
          // Recurse into the subRows
          groupedRows = groupedRows.map(function (rowGroup) {
            var subRows = groupRecursively(rowGroup[subRowsKey], keys, i + 1);
            return _extends({}, rowGroup, _defineProperty({}, subRowsKey, subRows), aggregate(subRows));
          });
          return groupedRows;
        };
        resolvedData = groupRecursively(resolvedData, pivotBy);
      })();
    }

    return _extends({}, newState, {
      resolvedData: resolvedData,
      pivotColumn: pivotColumn,
      allVisibleColumns: allVisibleColumns,
      headerGroups: headerGroups,
      allDecoratedColumns: allDecoratedColumns,
      hasHeaderGroups: hasHeaderGroups
    });
  },
  getSortedData: function getSortedData(resolvedState) {
    var manual = resolvedState.manual,
        sorting = resolvedState.sorting,
        filtering = resolvedState.filtering,
        showFilters = resolvedState.showFilters,
        defaultFilterMethod = resolvedState.defaultFilterMethod,
        resolvedData = resolvedState.resolvedData,
        allVisibleColumns = resolvedState.allVisibleColumns;

    // Resolve the data from either manual data or sorted data

    return {
      sortedData: manual ? resolvedData : this.sortData(this.filterData(resolvedData, showFilters, filtering, defaultFilterMethod, allVisibleColumns), sorting)
    };
  },
  fireOnChange: function fireOnChange() {
    this.props.onChange(this.getResolvedState(), this);
  },
  getPropOrState: function getPropOrState(key) {
    return _utils2.default.getFirstDefined(this.props[key], this.state[key]);
  },
  getStateOrProp: function getStateOrProp(key) {
    return _utils2.default.getFirstDefined(this.state[key], this.props[key]);
  },
  filterData: function filterData(data, showFilters, filtering, defaultFilterMethod, allVisibleColumns) {
    var _this2 = this;

    var filteredData = data;

    if (showFilters && filtering.length) {
      filteredData = filtering.reduce(function (filteredSoFar, nextFilter) {
        return filteredSoFar.filter(function (row) {
          var column = void 0;

          if (nextFilter.pivotId) {
            var parentColumn = allVisibleColumns.find(function (x) {
              return x.id === nextFilter.id;
            });
            column = parentColumn.pivotColumns.find(function (x) {
              return x.id === nextFilter.pivotId;
            });
          } else {
            column = allVisibleColumns.find(function (x) {
              return x.id === nextFilter.id;
            });
          }

          var filterMethod = column.filterMethod || defaultFilterMethod;

          return filterMethod(nextFilter, row, column);
        });
      }, filteredData);

      // Apply the filter to the subrows if we are pivoting, and then
      // filter any rows without subcolumns because it would be strange to show
      filteredData = filteredData.map(function (row) {
        if (!row[_this2.props.subRowsKey]) {
          return row;
        }
        return _extends({}, row, _defineProperty({}, _this2.props.subRowsKey, _this2.filterData(row[_this2.props.subRowsKey], showFilters, filtering, defaultFilterMethod, allVisibleColumns)));
      }).filter(function (row) {
        if (!row[_this2.props.subRowsKey]) {
          return true;
        }
        return row[_this2.props.subRowsKey].length > 0;
      });
    }

    return filteredData;
  },
  sortData: function sortData(data, sorting) {
    var _this3 = this;

    if (!sorting.length) {
      return data;
    }

    var sorted = _utils2.default.orderBy(data, sorting.map(function (sort) {
      return function (row) {
        if (row[sort.id] === null || row[sort.id] === undefined) {
          return -Infinity;
        }
        return typeof row[sort.id] === 'string' ? row[sort.id].toLowerCase() : row[sort.id];
      };
    }), sorting.map(function (d) {
      return !d.desc;
    }));

    return sorted.map(function (row) {
      if (!row[_this3.props.subRowsKey]) {
        return row;
      }
      return _extends({}, row, _defineProperty({}, _this3.props.subRowsKey, _this3.sortData(row[_this3.props.subRowsKey], sorting)));
    });
  },
  getMinRows: function getMinRows() {
    return _utils2.default.getFirstDefined(this.props.minRows, this.getStateOrProp('pageSize'));
  },


  // User actions
  onPageChange: function onPageChange(page) {
    var _this4 = this;

    var _props = this.props,
        onPageChange = _props.onPageChange,
        collapseOnPageChange = _props.collapseOnPageChange;

    if (onPageChange) {
      return onPageChange(page);
    }
    var newState = { page: page };
    if (collapseOnPageChange) {
      newState.expandedRows = {};
    }
    this.setStateWithData(newState, function () {
      _this4.fireOnChange();
    });
  },
  onPageSizeChange: function onPageSizeChange(newPageSize) {
    var _this5 = this;

    var onPageSizeChange = this.props.onPageSizeChange;

    var _getResolvedState = this.getResolvedState(),
        pageSize = _getResolvedState.pageSize,
        page = _getResolvedState.page;

    // Normalize the page to display


    var currentRow = pageSize * page;
    var newPage = Math.floor(currentRow / newPageSize);

    if (onPageSizeChange) {
      return onPageSizeChange(newPageSize, newPage);
    }

    this.setStateWithData({
      pageSize: newPageSize,
      page: newPage
    }, function () {
      _this5.fireOnChange();
    });
  },
  sortColumn: function sortColumn(column, additive) {
    var _this6 = this;

    var _getResolvedState2 = this.getResolvedState(),
        sorting = _getResolvedState2.sorting,
        skipNextSort = _getResolvedState2.skipNextSort;

    // we can't stop event propagation from the column resize move handlers
    // attached to the document because of react's synthetic events
    // so we have to prevent the sort function from actually sorting
    // if we click on the column resize element within a header.


    if (skipNextSort) {
      this.setStateWithData({
        skipNextSort: false
      });
      return;
    }

    var onSortingChange = this.props.onSortingChange;

    if (onSortingChange) {
      return onSortingChange(column, additive);
    }
    var newSorting = _utils2.default.clone(sorting || []).map(function (d) {
      d.desc = _utils2.default.isSortingDesc(d);
      return d;
    });
    if (!_utils2.default.isArray(column)) {
      // Single-Sort
      var existingIndex = newSorting.findIndex(function (d) {
        return d.id === column.id;
      });
      if (existingIndex > -1) {
        var existing = newSorting[existingIndex];
        if (existing.desc) {
          if (additive) {
            newSorting.splice(existingIndex, 1);
          } else {
            existing.desc = false;
            newSorting = [existing];
          }
        } else {
          existing.desc = true;
          if (!additive) {
            newSorting = [existing];
          }
        }
      } else {
        if (additive) {
          newSorting.push({
            id: column.id,
            desc: false
          });
        } else {
          newSorting = [{
            id: column.id,
            desc: false
          }];
        }
      }
    } else {
      (function () {
        // Multi-Sort
        var existingIndex = newSorting.findIndex(function (d) {
          return d.id === column[0].id;
        });
        // Existing Sorted Column
        if (existingIndex > -1) {
          var _existing = newSorting[existingIndex];
          if (_existing.desc) {
            if (additive) {
              newSorting.splice(existingIndex, column.length);
            } else {
              column.forEach(function (d, i) {
                newSorting[existingIndex + i].desc = false;
              });
            }
          } else {
            column.forEach(function (d, i) {
              newSorting[existingIndex + i].desc = true;
            });
          }
          if (!additive) {
            newSorting = newSorting.slice(existingIndex, column.length);
          }
        } else {
          // New Sort Column
          if (additive) {
            newSorting = newSorting.concat(column.map(function (d) {
              return {
                id: d.id,
                desc: false
              };
            }));
          } else {
            newSorting = column.map(function (d) {
              return {
                id: d.id,
                desc: false
              };
            });
          }
        }
      })();
    }
    this.setStateWithData({
      page: !sorting.length && newSorting.length || !additive ? 0 : this.state.page,
      sorting: newSorting
    }, function () {
      _this6.fireOnChange();
    });
  },
  filterColumn: function filterColumn(column, value, pivotColumn) {
    var _this7 = this;

    var _getResolvedState3 = this.getResolvedState(),
        filtering = _getResolvedState3.filtering;

    var onFilteringChange = this.props.onFilteringChange;


    if (onFilteringChange) {
      return onFilteringChange(column, value, pivotColumn);
    }

    // Remove old filter first if it exists
    var newFiltering = (filtering || []).filter(function (x) {
      if (x.id !== column.id) {
        return true;
      }
      if (x.pivotId) {
        if (pivotColumn) {
          return x.pivotId !== pivotColumn.id;
        }
        return true;
      }
    });

    if (value !== '') {
      newFiltering.push({
        id: column.id,
        value: value,
        pivotId: pivotColumn ? pivotColumn.id : undefined
      });
    }

    this.setStateWithData({
      filtering: newFiltering
    }, function () {
      _this7.fireOnChange();
    });
  },
  resizeColumnStart: function resizeColumnStart(column, event, isTouch) {
    var _this8 = this;

    var onResize = this.props.onResize;


    if (onResize) {
      return onResize(column, event, isTouch);
    }

    var parentWidth = event.target.parentElement.getBoundingClientRect().width;

    var pageX = void 0;
    if (isTouch) {
      pageX = event.changedTouches[0].pageX;
    } else {
      pageX = event.pageX;
    }

    this.setStateWithData({
      currentlyResizing: {
        id: column.id,
        startX: pageX,
        parentWidth: parentWidth
      }
    }, function () {
      if (isTouch) {
        document.addEventListener('touchmove', _this8.resizeColumnMoving);
        document.addEventListener('touchcancel', _this8.resizeColumnEnd);
        document.addEventListener('touchend', _this8.resizeColumnEnd);
      } else {
        document.addEventListener('mousemove', _this8.resizeColumnMoving);
        document.addEventListener('mouseup', _this8.resizeColumnEnd);
        document.addEventListener('mouseleave', _this8.resizeColumnEnd);
      }
    });
  },
  resizeColumnEnd: function resizeColumnEnd(event) {
    var isTouch = event.type === 'touchend' || event.type === 'touchcancel';

    if (isTouch) {
      document.removeEventListener('touchmove', this.resizeColumnMoving);
      document.removeEventListener('touchcancel', this.resizeColumnEnd);
      document.removeEventListener('touchend', this.resizeColumnEnd);
    }

    // If its a touch event clear the mouse one's as well because sometimes
    // the mouseDown event gets called as well, but the mouseUp event doesn't
    document.removeEventListener('mousemove', this.resizeColumnMoving);
    document.removeEventListener('mouseup', this.resizeColumnEnd);
    document.removeEventListener('mouseleave', this.resizeColumnEnd);

    // The touch events don't propagate up to the sorting's onMouseDown event so
    // no need to prevent it from happening or else the first click after a touch
    // event resize will not sort the column.
    if (!isTouch) {
      this.setStateWithData({
        skipNextSort: true
      });
    }
  },
  resizeColumnMoving: function resizeColumnMoving(event) {
    var _getResolvedState4 = this.getResolvedState(),
        resizing = _getResolvedState4.resizing,
        currentlyResizing = _getResolvedState4.currentlyResizing;

    // Delete old value


    var newResizing = resizing.filter(function (x) {
      return x.id !== currentlyResizing.id;
    });

    var pageX = void 0;

    if (event.type === 'touchmove') {
      pageX = event.changedTouches[0].pageX;
    } else if (event.type === 'mousemove') {
      pageX = event.pageX;
    }

    // Set the min size to 10 to account for margin and border or else the group headers don't line up correctly
    var newWidth = Math.max(currentlyResizing.parentWidth + pageX - currentlyResizing.startX, 11);

    newResizing.push({
      id: currentlyResizing.id,
      value: newWidth
    });

    this.setStateWithData({
      resizing: newResizing
    });
  }
};
